dist2src/worker/updater.py

```diff
@@ -53,67 +52,65 @@ def check_updates(
         else:
             logger.debug("Celery tasks created never expire")
 
-        m = re.match(
-            r"(?P<fork>forks\/)?((?P<owner>.+)\/)?(?P<namespace>.+)",
-            self.cfg.src_git_namespace,
+        src_gitlab_group = self.cfg.src_git_svc.gitlab_instance.groups.get(
+            self.cfg.src_git_namespace
         )
-
-        # Get repositories in source-git
-        url = f"{self.cfg.src_git_svc.api_url}projects"
-        params = {
-            "namespace": m["namespace"],
-            "pattern": project,
-            "fork": m["fork"] is not None,
-            "per_page": self.PROJECTS_PER_PAGE,
-            "owner": m["owner"] or None,
-            "short": True,
-        }
-        while url:
-            r = self.cfg.src_git_svc.call_api(url, params=params)
-            # The URL in 'next' already has the required parameters
-            url, params = r["pagination"]["next"], None
-            logger.debug(
-                f"Next page: {url!r}. Total pages: {r['pagination']['pages']!r}."
+        if project:
+            # update only the selected project
+            gitlab_project = self.cfg.src_git_svc.get_project(
+                repo=project, namespace=self.cfg.src_git_namespace
             )
-            for src_git_project in r["projects"]:
-                try:
-                    self._check_project(src_git_project["name"], branch)
-                except (OgrException, RetryError):
-                    logger.exception(
-                        f"Failed checking project {src_git_project['name']!r}"
+            self.check_and_update_project(gitlab_project, branch=branch)
+        else:
```

Comment:
Couldn't this `if...else` simplified by setting `projects` to `[project]` if `project` is defined or get the list of projects from the API if it isn't?

Something like:

```python
n = 1
projects = [project] if project else src_gitlab_group.projects.list(page=n, per_page=self.PROJECTS_PER_PAGE)
while projects:
    for project_ in projects:
    ...
    n += 1
    projects = src_gitlab_group.projects.list(page=n, per_page=self.PROJECTS_PER_PAGE)
```

Reply:
I don't see how your code is a simplification.

It actually does something else: it would start enumerating projects in the namespace even if `project` was set.

Please provide complete diff since it's hard to understand what you mean.

Reply:
Sorry, the last line should have been

```python
projects = [] if project else src_gitlab_group.projects.list(page=n, per_page=self.PROJECTS_PER_PAGE)
```

Reply:
Is this what you had in mind?

Sadly we cannot eliminate the first `if` because `project` is str while project_g is `SomethingGitlabGroupProject` and we wouldn't be able to get down into the for-loop. Though we could wrap that `self.cfg.src_git_svc.get_project` and `self.check_and_update_project` in a method to make the code-block shorter.

```python
        if project:
            # update only the selected project
            gitlab_project = self.cfg.src_git_svc.get_project(
                repo=project, namespace=self.cfg.src_git_namespace
            )
            self.check_and_update_project(gitlab_project, branch=branch)
        else:
            # enumerate all projects in the src/ GitLab group and try to update them
            n = 1
            projects = src_gitlab_group.projects.list(
                page=n, per_page=self.PROJECTS_PER_PAGE
            )
            while projects:
                for project_g in projects:
                    # we need to explicitly get every project from the API because those returned
                    # above are limited:
                    #   GroupProject objects returned by this API call are very limited,
                    #   and do not provide all the features of Project objects.
                    #   If you need to manipulate projects, create a new Project object.
                    # https://python-gitlab.readthedocs.io/en/latest/gl_objects/groups.html
                    gitlab_project = self.cfg.src_git_svc.get_project(
                        repo=project_g.name, namespace=self.cfg.src_git_namespace
                    )
                    self.check_and_update_project(gitlab_project, branch=branch)
                projects = src_gitlab_group.projects.list(
                    page=n, per_page=self.PROJECTS_PER_PAGE
                )
                n += 1
```

Reply:
Oh, sorry, I didn't notice that.

How about instead of passing a `GitlabProject` down to processing, pass down only the name of the project? That is the thing that is mostly used in calls branching out from `check_and_update_project()`.

A `GitlabProject` could be retrieved right before getting the tags in `_get_out_of_date_branches()`. If I'm not mistaken, it's not required before.

I'm thinking about doing something like [this](https://github.com/csomh/dist-git-to-source-git/commit/64c08deaed4f74a48ed48e8679fb1a79b2f93818) (untested code :exclamation:)